### ISSUE TITLE
Question about best practice for YOLO small object detection: high-resolution full-image training vs SAHI sliced training/inference

### ISSUE THREAD

**Issue by Lan1tan** [ultralytics/ultralytics#24732]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hi Ultralytics team,

I have a question about the best practice for small object detection with YOLO and SAHI.

My dataset consists of high-resolution images, for example around 3000x3000. The objects I need to detect are relatively small in the original images.

I understand that if I train a YOLO model with imgsz=640 on full images, the original image will be resized to 640, so small objects may become very tiny in the model input. SAHI sliced inference can help because it crops the large image into smaller patches, such as 640x640, and runs detection on each slice. In this way, the small objects can keep more pixels in the model input.

I would like to ask for guidance on the following questions.

1. Should training also use sliced images if inference uses SAHI?

If I plan to use SAHI during inference, is it recommended to also train the model on sliced images?

For example:

Original image: 3000x3000

Training:
- Slice original images into 640x640 patches
- Use overlap=0.2
- Convert the original annotations into sliced-image annotations
- Train YOLO with imgsz=640

Inference:
- Use SAHI with slice_width=640
- Use SAHI with slice_height=640
- Use overlap=0.2

Would this be better than training on full images with imgsz=640 and only using SAHI during inference?

In other words, for small object detection, should the training data distribution match the SAHI inference distribution?

2. How does full-image high-resolution training compare with SAHI sliced training/inference?

If GPU memory and computation are not a problem, how would full-image high-resolution training and inference compare with SAHI sliced training/inference?

For example:

Option A:
- Train on the full 3000x3000 image with imgsz=3000
- Infer on the full 3000x3000 image with imgsz=3000

Option B:
- Slice the original 3000x3000 image into 640x640 patches
- Train YOLO with imgsz=640
- Infer with SAHI using 640x640 slices

From my understanding, in both cases, a small object in the original image would have a similar pixel size in the model input.

For example, if an object is 20x20 pixels in the original image:

Option A:
input_region=3000, imgsz=3000
object size in model input ≈ 20 * 3000 / 3000 = 20 pixels

Option B:
input_region=640, imgsz=640
object size in model input ≈ 20 * 640 / 640 = 20 pixels

So theoretically, is full-image high-resolution training comparable to SAHI sliced training/inference?

Could full-image high-resolution training even be better because it preserves global context and avoids slicing boundary issues?

Or is SAHI still preferred in practice because it is more memory-efficient, easier to train, and more practical for deployment?

3. Should the SAHI inference slice size match the training slice size?

If I use sliced training, should the SAHI inference slice size be the same as the training slice size?

For example, if I train with:

slice_size=640
imgsz=640
overlap=0.2

Should I also use similar settings during SAHI inference?

slice_width=640
slice_height=640
overlap=0.2

My current understanding is that the object scale seen by the model depends on this relationship:

object size in model input ≈ original object size × imgsz / slice_size

So if the training uses slice_size=640 and imgsz=640, but inference uses a very different slice size, the object scale distribution may change.

Is this understanding correct?

4. Can smaller slices improve small object detection?

If the goal is to make small objects appear larger in the model input, would it make sense to use a smaller slice size while keeping imgsz fixed?

For example:

Original object size: 20x20 pixels

slice_size=640, imgsz=640
=> object size in model input ≈ 20x20 pixels

slice_size=320, imgsz=640
=> object size in model input ≈ 40x40 pixels

Based on this, smaller slices may make small objects easier to detect.

However, I also understand that smaller slices may have disadvantages:

- Less surrounding context
- More slices per image
- Slower inference
- More duplicated detections
- More boundary cases
- More complicated merging/NMS

Is this a reasonable trade-off to test for small object detection?

5. What is the recommended way to prepare annotations for sliced training?

If using sliced training, what is the recommended annotation workflow?

Should I annotate the original full-resolution images first and then programmatically convert the bounding boxes into sliced-image annotations?

For example:

Original full image annotation
        ↓
Slice image into patches
        ↓
Convert bbox coordinates into slice-local coordinates
        ↓
Clip boxes at slice boundaries
        ↓
Filter boxes that are too heavily truncated
        ↓
Train YOLO on sliced images

Or is it better to slice the images first and then manually annotate each slice?

My current assumption is that annotating the original full-resolution images first is better, because it avoids duplicated labels caused by overlapping slices and makes it easier to regenerate sliced datasets with different slice sizes or overlap ratios.

Is this the recommended approach?

Summary of my current understanding

My current understanding is:

object size in model input ≈ original object size × imgsz / input_region_size

For full-image training:

input_region_size = original image size

For SAHI sliced inference or sliced training:

input_region_size = slice_size

Therefore:

Full image 3000x3000 with imgsz=640:
20px object => about 4.3px in model input

SAHI slice 640x640 with imgsz=640:
20px object => about 20px in model input

SAHI slice 320x320 with imgsz=640:
20px object => about 40px in model input

Full image 3000x3000 with imgsz=3000:
20px object => about 20px in model input

So for small object detection on high-resolution images, I am trying to understand which strategy is generally recommended:

A. Full-image training with imgsz=640 + SAHI inference
B. Sliced training with slice_size=640, imgsz=640 + SAHI inference
C. Full-image high-resolution training/inference with imgsz=3000
D. Smaller sliced training, for example slice_size=320, imgsz=640 + SAHI inference

Could you please provide guidance on which approach is recommended in practice for YOLO small object detection on high-resolution images?

Thanks!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Lan1tan, thank you for your interest in Ultralytics 🚀! This is an automated response to help get you started, and an Ultralytics engineer will also assist you soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples and where many of the most common questions may already be answered.

Since this is a custom training and small-object detection ❓ question, please provide as much information as possible, including dataset image examples, sample labels, your current training configuration, validation results, and any training logs you have already collected. It would also help to confirm whether you have already tested any of the strategies you listed and how they compared. Please also verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips).

If you encounter any unexpected behavior while testing these setups, please also share a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) so we can investigate more effectively 🔍

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
Hello.

1. Training should use sliced images same as inference.
2. Using full image is better because it avoids issues with false positives at the edges of slices.
3. Yes
4. Depends on the type of object. But an object of 40x40 in a slice of 640x640 already has a very large surrounding context so it shouldn't have issue with context.
5. Generating slices from full resolution images is better because it lets you regenerate slices like you mentioned.

**Comment by MUKILAN0608** [comment]
Thanks for the detailed write-up — you've already reasoned through most of the important trade-offs correctly.

A useful way to think about this is that YOLO does not care about the *absolute size* of an object in the original image. What matters is how large that object appears in the image that is actually fed into the model.

Your formula is a good approximation:

**effective object size ≈ original object size × imgsz / input region size**

where the input region is either the full image or a SAHI tile.

For example, if a 20×20 px object exists in a 3000×3000 image:

* Full-image training with `imgsz=640` shrinks that object to roughly 4 px.
* A 640×640 tile with `imgsz=640` keeps it around 20 px.
* A 320×320 tile with `imgsz=640` makes it appear around 40 px.

This is exactly why SAHI often helps with small objects: it is not doing anything magical, it is simply preventing small objects from being aggressively downscaled.

### Should training use slices if inference uses SAHI?

In most small-object scenarios, yes.

If the model is trained on full 3000×3000 images resized to 640, it mainly learns from objects that appear very small in the input.

Then at inference time, SAHI suddenly presents those same objects much larger because each tile is resized independently. The model can still benefit from SAHI, but there is now a train/inference scale mismatch.

Training on slices and inferring on slices usually produces a more consistent setup because the model sees roughly the same object scales during both phases.

I would generally recommend:

* Annotate full-resolution images
* Generate training tiles
* Train on tiles
* Use SAHI during inference with a similar tile size

for most high-resolution small-object datasets.

### What about training on full 3000×3000 images?

Your comparison between:

**Option A**

* Train/infer on full 3000×3000
* `imgsz=3000`

and

**Option B**

* Train on 640×640 tiles
* Infer with SAHI 640×640

is also largely correct.

In both cases, the object may occupy roughly the same number of pixels in the model input.

The difference is that Option A preserves the entire scene context while Option B only shows a local region.

If GPU memory, training speed, batch size, and deployment constraints were completely irrelevant, then full-resolution training could absolutely be competitive and in some datasets even better because the model retains global context and avoids tile-boundary issues.

The reason tiled workflows are so common is mostly practical:

* Lower memory usage
* Larger batch sizes
* Faster training
* Easier deployment
* Works well on extremely large imagery

So I would not say SAHI is inherently better than full-resolution training. It is often the more practical way to achieve a similar effective object scale without needing enormous compute.

### Should training and inference tile sizes match?

As a starting point, yes.

If you train on:

* tile size = 640
* imgsz = 640

then using approximately the same tile size during SAHI inference is usually the safest choice.

The model learns a particular object-scale distribution during training. If you suddenly switch to 320 px tiles at inference, objects will appear much larger than what the model saw during training.

That does not automatically hurt performance, but it introduces a distribution shift.

My recommendation would be:

1. Train on 640 tiles.
2. Infer with 640 tiles first.
3. Then experiment with smaller inference tiles and compare mAP/recall.

Let the validation metrics decide.

### Can smaller tiles help?

Definitely.

Your reasoning here is correct.

Smaller tiles effectively zoom in on objects:

* 640 tile → 20 px object stays ~20 px
* 320 tile → same object appears ~40 px

This can improve detection of extremely small targets.

The trade-off is exactly what you listed:

* Less context
* More tiles per image
* More overlap processing
* More duplicate detections
* More boundary cases
* Slower inference

So smaller tiles are not universally better. They are often worth testing when objects are extremely tiny relative to the original image.

### Annotation workflow

I strongly recommend annotating the original full-resolution images first and generating tiled annotations programmatically afterward.

Something like:

Full image annotation
→ Generate tiles
→ Convert coordinates
→ Clip boundary-crossing boxes
→ Optionally remove heavily truncated boxes
→ Train

This gives you a single source of truth and lets you regenerate datasets with different tile sizes or overlap values whenever you want.

Manually annotating tiles usually becomes painful because overlapping tiles create duplicated labels and make future experiments much harder.

### My practical recommendation

For a dataset of ~3000×3000 images containing genuinely small objects, I would start with:

* Annotate full images
* Generate 640×640 training tiles (with overlap if needed)
* Train YOLO at `imgsz=640`
* Use SAHI inference with roughly the same tile size
* Evaluate
* If objects are still very small, experiment with smaller tiles such as 512 or 320

That is the workflow I would expect to work well in most real-world small-object detection projects.

**Comment by glenn-jocher** [comment]
That’s the right direction: for 3000 x 3000 imagery with small targets, I’d start with full-image annotations as the source of truth, generate `640 x 640` training tiles, train at `imgsz=640`, and use similar SAHI tile sizes first, then only test smaller tiles if validation improves small-object recall; the [SAHI sliced inference guide](https://docs.ultralytics.com/guides/sahi-tiled-inference) is the best baseline to follow.
